## Identifying Opportunities to Reduce Car Crashes in Chicago 

### Book 1: Prepare the dataset for exploration and modelling

### Business Case

The Vehicle Safety Board is looking to reduce the number of accidents in the City of Chicago. Utilizing the car crash data available in Chicago Data Portal, we will perform some meaningful EDA and provide recommendations to the Vehicle Safety Board. 

Firstly, we will do some data exploration to answer the three main problem statements. We will then use the results to identify areas for improvement and come up with actionable steps that will help reduce the number of accidents in the City of Chicago.

**1. Identify high-density areas of car crash in Chicago**

**2. Analyze control failures to identify opportunities for improvement**

**3. Check for trends in the time of crash to relocate resources appropriately**

Secondly, we will create a ML classifier to categorize accidents in four main cateogories:

1. Injury & Damages over 1,500 dollars
2. Injury & Damages under 1,500 dollars 
3. No Injury & Damages over 1,500 dollars
4. No Injury & Damages under 1,500 dollars

Using these four classes, we will predict what action, event or conditions are most likely to lead to an injury and/or damages in either of the four classes.

### Dataset

Chicago Car Crashes https://data.cityofchicago.org/Transportation/Traffic-Crashes-Crashes/85ca-t3if

Vehicle Data https://data.cityofchicago.org/Transportation/Traffic-Crashes-Vehicles/68nd-jvt3

Driver/Passenger Data https://data.cityofchicago.org/Transportation/Traffic-Crashes-People/u6pd-qa9d

### Objective

- Inspect all datasets and explore it's contents.
- Clean each dataset by removing missing values and deleting columns irrelevant to our business case.
- Merge the three datasets to create a master dataset.
- Explore each variable and perform feature engineering to prepare the data for modeling.

### Data Understanding

In [1]:
# Import relevant packages

import pandas as pd
import numpy as np

In [2]:
# Import all datasets

df_crashes_raw = pd.read_csv(r'Datasets/Traffic_Crashes_-_Crashes.csv')
df_vehicles_raw = pd.read_csv(r'Datasets/Traffic_Crashes_-_Vehicles.csv', dtype='unicode')
df_people_raw = pd.read_csv(r'Datasets/Traffic_Crashes_-_People.csv', dtype='unicode')

### Traffic Crashes Dataset

In [3]:
df_crashes_raw.head()

CRASH_RECORD_ID     RD_NO  \
0  4fd0a3e0897b3335b94cd8d5b2d2b350eb691add56c62d...  JC343143   
1  009e9e67203442370272e1a13d6ee51a4155dac65e583d...  JA329216   
2  ee9283eff3a55ac50ee58f3d9528ce1d689b1c4180b4c4...  JD292400   
3  f8960f698e870ebdc60b521b2a141a5395556bc3704191...  JD293602   
4  8eaa2678d1a127804ee9b8c35ddf7d63d913c14eda61d6...  JD290451   

  CRASH_DATE_EST_I              CRASH_DATE  POSTED_SPEED_LIMIT  \
0              NaN  07/10/2019 05:56:00 PM                  35   
1              NaN  06/30/2017 04:00:00 PM                  35   
2              NaN  07/10/2020 10:25:00 AM                  30   
3              NaN  07/11/2020 01:00:00 AM                  30   
4              NaN  07/08/2020 02:00:00 PM                  20   

  TRAFFIC_CONTROL_DEVICE      DEVICE_CONDITION WEATHER_CONDITION  \
0            NO CONTROLS           NO CONTROLS             CLEAR   
1      STOP SIGN/FLASHER  FUNCTIONING PROPERLY             CLEAR   
2         TRAFFIC SIGNAL  FUNCTIONING PROPERLY             CLEAR   
3            NO CONTROLS           NO CONTROLS             CLEAR   
4            NO CONTROLS           NO CONTROLS             CLEAR   

  LIGHTING_CONDITION      FIRST_CRASH_TYPE  ... INJURIES_NON_INCAPACITATING  \
0           DAYLIGHT               TURNING  ...                         0.0   
1           DAYLIGHT               TURNING  ...                         0.0   
2           DAYLIGHT              REAR END  ...                         0.0   
3           DARKNESS  PARKED MOTOR VEHICLE  ...                         0.0   
4           DAYLIGHT  PARKED MOTOR VEHICLE  ...                         0.0   

   INJURIES_REPORTED_NOT_EVIDENT INJURIES_NO_INDICATION INJURIES_UNKNOWN  \
0                            0.0                    3.0              0.0   
1                            0.0                    3.0              0.0   
2                            0.0                    3.0              0.0   
3                            0.0                    3.0              0.0   
4                            0.0                    1.0              0.0   

  CRASH_HOUR CRASH_DAY_OF_WEEK CRASH_MONTH   LATITUDE  LONGITUDE  \
0         17                 4           7  41.919664 -87.773288   
1         16                 6           6  41.741804 -87.740954   
2         10                 6           7  41.773456 -87.585022   
3          1                 7           7  41.802119 -87.622115   
4         14                 4           7        NaN        NaN   

                                   LOCATION  
0  POINT (-87.773287883007 41.919663832993)  
1  POINT (-87.740953581987 41.741803598989)  
2  POINT (-87.585022352022 41.773455972008)  
3  POINT (-87.622114914961 41.802118543011)  
4                                       NaN  

[5 rows x 49 columns]

In [4]:
# Check all columns in the dataset

df_crashes_raw.columns

Index(['CRASH_RECORD_ID', 'RD_NO', 'CRASH_DATE_EST_I', 'CRASH_DATE',
       'POSTED_SPEED_LIMIT', 'TRAFFIC_CONTROL_DEVICE', 'DEVICE_CONDITION',
       'WEATHER_CONDITION', 'LIGHTING_CONDITION', 'FIRST_CRASH_TYPE',
       'TRAFFICWAY_TYPE', 'LANE_CNT', 'ALIGNMENT', 'ROADWAY_SURFACE_COND',
       'ROAD_DEFECT', 'REPORT_TYPE', 'CRASH_TYPE', 'INTERSECTION_RELATED_I',
       'NOT_RIGHT_OF_WAY_I', 'HIT_AND_RUN_I', 'DAMAGE', 'DATE_POLICE_NOTIFIED',
       'PRIM_CONTRIBUTORY_CAUSE', 'SEC_CONTRIBUTORY_CAUSE', 'STREET_NO',
       'STREET_DIRECTION', 'STREET_NAME', 'BEAT_OF_OCCURRENCE',
       'PHOTOS_TAKEN_I', 'STATEMENTS_TAKEN_I', 'DOORING_I', 'WORK_ZONE_I',
       'WORK_ZONE_TYPE', 'WORKERS_PRESENT_I', 'NUM_UNITS',
       'MOST_SEVERE_INJURY', 'INJURIES_TOTAL', 'INJURIES_FATAL',
       'INJURIES_INCAPACITATING', 'INJURIES_NON_INCAPACITATING',
       'INJURIES_REPORTED_NOT_EVIDENT', 'INJURIES_NO_INDICATION',
       'INJURIES_UNKNOWN', 'CRASH_HOUR', 'CRASH_DAY_OF_WEEK', 'CRASH_MONTH',
       'LA

The description for each column in the datasets are available in the Chicago Data Portal.

Chicago Car Crashes https://data.cityofchicago.org/Transportation/Traffic-Crashes-Crashes/85ca-t3if

Vehicle Data https://data.cityofchicago.org/Transportation/Traffic-Crashes-Vehicles/68nd-jvt3

Driver/Passenger Data https://data.cityofchicago.org/Transportation/Traffic-Crashes-People/u6pd-qa9d

For the purpose of this project, we have decided to remove the following columns.

In [5]:
columns_to_delete = ['RD_NO', 'CRASH_DATE_EST_I', 'CRASH_DATE','LANE_CNT','REPORT_TYPE','DATE_POLICE_NOTIFIED', 
                     'SEC_CONTRIBUTORY_CAUSE','STREET_NO','STREET_DIRECTION', 'STREET_NAME','PHOTOS_TAKEN_I', 
                     'STATEMENTS_TAKEN_I', 'BEAT_OF_OCCURRENCE', 'MOST_SEVERE_INJURY', 'INJURIES_TOTAL',
                     'INJURIES_FATAL','INJURIES_INCAPACITATING', 'INJURIES_NON_INCAPACITATING', 
                     'INJURIES_REPORTED_NOT_EVIDENT', 'INJURIES_NO_INDICATION', 'INJURIES_UNKNOWN']

In [6]:
df_crashes_raw[:]

CRASH_RECORD_ID     RD_NO  \
0       4fd0a3e0897b3335b94cd8d5b2d2b350eb691add56c62d...  JC343143   
1       009e9e67203442370272e1a13d6ee51a4155dac65e583d...  JA329216   
2       ee9283eff3a55ac50ee58f3d9528ce1d689b1c4180b4c4...  JD292400   
3       f8960f698e870ebdc60b521b2a141a5395556bc3704191...  JD293602   
4       8eaa2678d1a127804ee9b8c35ddf7d63d913c14eda61d6...  JD290451   
...                                                   ...       ...   
555214  9a9db62f3334a1fad706f97c5a4ebb8485668447c176e2...  JE299347   
555215  db31327d28803316b8f44f0ec86d6e76a248934f3d1bfc...  JE295652   
555216  d51aae396db49981c7ee26ceb54dfcab3c4b06d0cc5d7d...  JE298826   
555217  6f9abc7e7f54095cef0fa17e16e6f72eb14f8d17d3b572...  JE300245   
555218  bc361965fe506c50f7899493c79b1f7d011376f753acaa...  JE300060   

       CRASH_DATE_EST_I              CRASH_DATE  POSTED_SPEED_LIMIT  \
0                   NaN  07/10/2019 05:56:00 PM                  35   
1                   NaN  06/30/2017 04:00:00 PM                  35   
2                   NaN  07/10/2020 10:25:00 AM                  30   
3                   NaN  07/11/2020 01:00:00 AM                  30   
4                   NaN  07/08/2020 02:00:00 PM                  20   
...                 ...                     ...                 ...   
555214              NaN  07/12/2021 06:37:00 PM                  10   
555215              NaN  07/07/2021 09:25:00 PM                  30   
555216              NaN  07/12/2021 10:00:00 AM                  30   
555217              NaN  07/13/2021 04:15:00 PM                  15   
555218              NaN  07/13/2021 11:30:00 AM                  30   

       TRAFFIC_CONTROL_DEVICE      DEVICE_CONDITION WEATHER_CONDITION  \
0                 NO CONTROLS           NO CONTROLS             CLEAR   
1           STOP SIGN/FLASHER  FUNCTIONING PROPERLY             CLEAR   
2              TRAFFIC SIGNAL  FUNCTIONING PROPERLY             CLEAR   
3                 NO CONTROLS           NO CONTROLS             CLEAR   
4                 NO CONTROLS           NO CONTROLS             CLEAR   
...                       ...                   ...               ...   
555214            NO CONTROLS           NO CONTROLS             CLEAR   
555215         TRAFFIC SIGNAL  FUNCTIONING PROPERLY           UNKNOWN   
555216            NO CONTROLS           NO CONTROLS             CLEAR   
555217            NO CONTROLS           NO CONTROLS             CLEAR   
555218            NO CONTROLS           NO CONTROLS             CLEAR   

       LIGHTING_CONDITION          FIRST_CRASH_TYPE  ...  \
0                DAYLIGHT                   TURNING  ...   
1                DAYLIGHT                   TURNING  ...   
2                DAYLIGHT                  REAR END  ...   
3                DARKNESS      PARKED MOTOR VEHICLE  ...   
4                DAYLIGHT      PARKED MOTOR VEHICLE  ...   
...                   ...                       ...  ...   
555214           DAYLIGHT      PARKED MOTOR VEHICLE  ...   
555215            UNKNOWN  SIDESWIPE SAME DIRECTION  ...   
555216           DAYLIGHT              REAR TO SIDE  ...   
555217           DAYLIGHT      PARKED MOTOR VEHICLE  ...   
555218           DAYLIGHT  SIDESWIPE SAME DIRECTION  ...   

       INJURIES_NON_INCAPACITATING  INJURIES_REPORTED_NOT_EVIDENT  \
0                              0.0                            0.0   
1                              0.0                            0.0   
2                              0.0                            0.0   
3                              0.0                            0.0   
4                              0.0                            0.0   
...                            ...                            ...   
555214                         0.0                            0.0   
555215                         0.0                            0.0   
555216                         0.0                            1.0   
555217                         0.0              

In [7]:
# Drop the irrelavant columns

df_crashes_clean = df_crashes_raw.drop(columns_to_delete, axis=1).copy()

We will now go into looking for missing or null values and remove them as necessary.

In [8]:
# Check for missing values

df_crashes_clean.isnull().sum()

CRASH_RECORD_ID                 0
POSTED_SPEED_LIMIT              0
TRAFFIC_CONTROL_DEVICE          0
DEVICE_CONDITION                0
WEATHER_CONDITION               0
LIGHTING_CONDITION              0
FIRST_CRASH_TYPE                0
TRAFFICWAY_TYPE                 0
ALIGNMENT                       0
ROADWAY_SURFACE_COND            0
ROAD_DEFECT                     0
CRASH_TYPE                      0
INTERSECTION_RELATED_I     429100
NOT_RIGHT_OF_WAY_I         528971
HIT_AND_RUN_I              387461
DAMAGE                          0
PRIM_CONTRIBUTORY_CAUSE         0
DOORING_I                  553422
WORK_ZONE_I                551731
WORK_ZONE_TYPE             552459
WORKERS_PRESENT_I          554350
NUM_UNITS                       0
CRASH_HOUR                      0
CRASH_DAY_OF_WEEK               0
CRASH_MONTH                     0
LATITUDE                     3207
LONGITUDE                    3207
LOCATION                     3207
dtype: int64

In [9]:
# Check the total count of the dataset

df_crashes_clean.shape

(555219, 28)

Out of 555,219 total values, some of the columns above show majority of missing values. We will drop these columns.

In [10]:
df_crashes_clean.drop(['INTERSECTION_RELATED_I', 'NOT_RIGHT_OF_WAY_I','HIT_AND_RUN_I',
                       'DOORING_I','WORK_ZONE_I','WORK_ZONE_TYPE', 'WORKERS_PRESENT_I'], axis=1, inplace=True)

In [11]:
df_crashes_clean.isnull().sum()

CRASH_RECORD_ID               0
POSTED_SPEED_LIMIT            0
TRAFFIC_CONTROL_DEVICE        0
DEVICE_CONDITION              0
WEATHER_CONDITION             0
LIGHTING_CONDITION            0
FIRST_CRASH_TYPE              0
TRAFFICWAY_TYPE               0
ALIGNMENT                     0
ROADWAY_SURFACE_COND          0
ROAD_DEFECT                   0
CRASH_TYPE                    0
DAMAGE                        0
PRIM_CONTRIBUTORY_CAUSE       0
NUM_UNITS                     0
CRASH_HOUR                    0
CRASH_DAY_OF_WEEK             0
CRASH_MONTH                   0
LATITUDE                   3207
LONGITUDE                  3207
LOCATION                   3207
dtype: int64

The missing values in the last three columns represent a very small portion of the data. To preserve information available in rest of the columns, we will replace these values with zero.

In [12]:
# Replace missing values with 0

df_crashes_clean.fillna(0, axis=0, inplace=True)

In [13]:
df_crashes_clean.isnull().sum()

CRASH_RECORD_ID            0
POSTED_SPEED_LIMIT         0
TRAFFIC_CONTROL_DEVICE     0
DEVICE_CONDITION           0
WEATHER_CONDITION          0
LIGHTING_CONDITION         0
FIRST_CRASH_TYPE           0
TRAFFICWAY_TYPE            0
ALIGNMENT                  0
ROADWAY_SURFACE_COND       0
ROAD_DEFECT                0
CRASH_TYPE                 0
DAMAGE                     0
PRIM_CONTRIBUTORY_CAUSE    0
NUM_UNITS                  0
CRASH_HOUR                 0
CRASH_DAY_OF_WEEK          0
CRASH_MONTH                0
LATITUDE                   0
LONGITUDE                  0
LOCATION                   0
dtype: int64

Let's explore other datasets to see what we can find.

### Traffic Vehicle Dataset

In [14]:
df_vehicles_raw.head()

CRASH_UNIT_ID                                    CRASH_RECORD_ID     RD_NO  \
0        829999  24ddf9fd8542199d832e1c223cc474e5601b356f1d77a6...  JD124535   
1        749947  81dc0de2ed92aa62baccab641fa377be7feb1cc47e6554...  JC451435   
2        749949  81dc0de2ed92aa62baccab641fa377be7feb1cc47e6554...  JC451435   
3        749950  81dc0de2ed92aa62baccab641fa377be7feb1cc47e6554...  JC451435   
4        871921  af84fb5c8d996fcd3aefd36593c3a02e6e7509eeb27568...  JD208731   

               CRASH_DATE UNIT_NO UNIT_TYPE NUM_PASSENGERS VEHICLE_ID  \
0  01/22/2020 06:25:00 AM       1    DRIVER            NaN     796949   
1  09/28/2019 03:30:00 AM       1    DRIVER            NaN     834816   
2  09/28/2019 03:30:00 AM       2    PARKED            NaN     834819   
3  09/28/2019 03:30:00 AM       3    PARKED            NaN     834817   
4  04/13/2020 10:50:00 PM       2    DRIVER            NaN     827212   

  CMRC_VEH_I                              MAKE  ... TRAILER1_LENGTH  \
0        NaN                          INFINITI  ...             NaN   
1        NaN                             HONDA  ...             NaN   
2        NaN                            TOYOTA  ...             NaN   
3        NaN  GENERAL MOTORS CORPORATION (GMC)  ...             NaN   
4        NaN                             BUICK  ...             NaN   

  TRAILER2_LENGTH TOTAL_VEHICLE_LENGTH AXLE_CNT VEHICLE_CONFIG  \
0             NaN                  NaN      NaN            NaN   
1             NaN                  NaN      NaN            NaN   
2             NaN                  NaN      NaN            NaN   
3             NaN                  NaN      NaN            NaN   
4             NaN                  NaN      NaN            NaN   

  CARGO_BODY_TYPE LOAD_TYPE HAZMAT_OUT_OF_SERVICE_I MCS_OUT_OF_SERVICE_I  \
0             NaN       NaN                     NaN                  NaN   
1             NaN       NaN                     NaN                  NaN   
2             NaN       NaN                     NaN                  NaN   
3             NaN       NaN                     NaN                  NaN   
4             NaN       NaN                     NaN                  NaN   

  HAZMAT_CLASS  
0          NaN  
1          NaN  
2          NaN  
3          NaN  
4          NaN  

[5 rows x 72 columns]

In [15]:
# Check all variables in the dataset

df_vehicles_raw.columns

Index(['CRASH_UNIT_ID', 'CRASH_RECORD_ID', 'RD_NO', 'CRASH_DATE', 'UNIT_NO',
       'UNIT_TYPE', 'NUM_PASSENGERS', 'VEHICLE_ID', 'CMRC_VEH_I', 'MAKE',
       'MODEL', 'LIC_PLATE_STATE', 'VEHICLE_YEAR', 'VEHICLE_DEFECT',
       'VEHICLE_TYPE', 'VEHICLE_USE', 'TRAVEL_DIRECTION', 'MANEUVER',
       'TOWED_I', 'FIRE_I', 'OCCUPANT_CNT', 'EXCEED_SPEED_LIMIT_I', 'TOWED_BY',
       'TOWED_TO', 'AREA_00_I', 'AREA_01_I', 'AREA_02_I', 'AREA_03_I',
       'AREA_04_I', 'AREA_05_I', 'AREA_06_I', 'AREA_07_I', 'AREA_08_I',
       'AREA_09_I', 'AREA_10_I', 'AREA_11_I', 'AREA_12_I', 'AREA_99_I',
       'FIRST_CONTACT_POINT', 'CMV_ID', 'USDOT_NO', 'CCMC_NO', 'ILCC_NO',
       'COMMERCIAL_SRC', 'GVWR', 'CARRIER_NAME', 'CARRIER_STATE',
       'CARRIER_CITY', 'HAZMAT_PLACARDS_I', 'HAZMAT_NAME', 'UN_NO',
       'HAZMAT_PRESENT_I', 'HAZMAT_REPORT_I', 'HAZMAT_REPORT_NO',
       'MCS_REPORT_I', 'MCS_REPORT_NO', 'HAZMAT_VIO_CAUSE_CRASH_I',
       'MCS_VIO_CAUSE_CRASH_I', 'IDOT_PERMIT_NO', 'WIDE_LOAD_I',
       '

Here also, we have a lot of variables that we do not need for our analysis. We will remove these columns. 

In [16]:
columns_to_delete_vehicles = ['CRASH_UNIT_ID', 'CRASH_RECORD_ID', 'RD_NO', 'CRASH_DATE',
       'NUM_PASSENGERS', 'CMRC_VEH_I', 'MAKE','MODEL', 'LIC_PLATE_STATE', 
       'VEHICLE_YEAR','TOWED_I', 'FIRE_I', 'OCCUPANT_CNT', 
       'TOWED_BY', 'TOWED_TO', 'AREA_00_I', 'AREA_01_I', 'AREA_02_I', 'AREA_03_I',
       'AREA_04_I', 'AREA_05_I', 'AREA_06_I', 'AREA_07_I', 'AREA_08_I',
       'AREA_09_I', 'AREA_10_I', 'AREA_11_I', 'AREA_12_I', 'AREA_99_I',
       'CMV_ID', 'USDOT_NO', 'CCMC_NO', 'ILCC_NO',
       'COMMERCIAL_SRC', 'GVWR', 'CARRIER_NAME', 'CARRIER_STATE',
       'CARRIER_CITY', 'HAZMAT_PLACARDS_I', 'HAZMAT_NAME', 'UN_NO',
       'HAZMAT_PRESENT_I', 'HAZMAT_REPORT_I', 'HAZMAT_REPORT_NO',
       'MCS_REPORT_I', 'MCS_REPORT_NO', 'HAZMAT_VIO_CAUSE_CRASH_I',
       'MCS_VIO_CAUSE_CRASH_I', 'IDOT_PERMIT_NO', 'WIDE_LOAD_I',
       'TRAILER1_WIDTH', 'TRAILER2_WIDTH', 'TRAILER1_LENGTH',
       'TRAILER2_LENGTH', 'TOTAL_VEHICLE_LENGTH', 'AXLE_CNT', 'VEHICLE_CONFIG',
       'CARGO_BODY_TYPE', 'LOAD_TYPE', 'HAZMAT_OUT_OF_SERVICE_I',
       'MCS_OUT_OF_SERVICE_I', 'HAZMAT_CLASS']

In [17]:
# Drop the irrelevant columns

df_vehicles_clean = df_vehicles_raw.drop(columns_to_delete_vehicles, axis=1).copy()

In [18]:
# Check for missing values

df_vehicles_clean.isnull().sum()

UNIT_NO                       0
UNIT_TYPE                  1626
VEHICLE_ID                26058
VEHICLE_DEFECT            26058
VEHICLE_TYPE              26058
VEHICLE_USE               26058
TRAVEL_DIRECTION          26058
MANEUVER                  26058
EXCEED_SPEED_LIMIT_I    1134303
FIRST_CONTACT_POINT       34702
dtype: int64

In [19]:
# Check the total count of the dataset

df_vehicles_clean.shape

(1136696, 10)

EXCEED_SPEED_LIMIT_I column has majority values missings. Let's drop this column first.

In [20]:
# Drop Exceed_speed_limit

df_vehicles_clean.drop('EXCEED_SPEED_LIMIT_I', axis=1, inplace=True)

In [21]:
df_vehicles_clean.isnull().sum()

UNIT_NO                    0
UNIT_TYPE               1626
VEHICLE_ID             26058
VEHICLE_DEFECT         26058
VEHICLE_TYPE           26058
VEHICLE_USE            26058
TRAVEL_DIRECTION       26058
MANEUVER               26058
FIRST_CONTACT_POINT    34702
dtype: int64

The rest of the missing values account for only 2% of the dataset. We will remove these values.

In [22]:
# Drop missing values

df_vehicles_clean.dropna(axis=0, inplace=True)

In [23]:
df_vehicles_clean.isnull().sum()

UNIT_NO                0
UNIT_TYPE              0
VEHICLE_ID             0
VEHICLE_DEFECT         0
VEHICLE_TYPE           0
VEHICLE_USE            0
TRAVEL_DIRECTION       0
MANEUVER               0
FIRST_CONTACT_POINT    0
dtype: int64

Moving on to people data set.

### Traffic People Dataset

In [24]:
df_people_raw.head()

PERSON_ID PERSON_TYPE                                    CRASH_RECORD_ID  \
0   O749947      DRIVER  81dc0de2ed92aa62baccab641fa377be7feb1cc47e6554...   
1   O871921      DRIVER  af84fb5c8d996fcd3aefd36593c3a02e6e7509eeb27568...   
2    O10018      DRIVER  71162af7bf22799b776547132ebf134b5b438dcf3dac6b...   
3    O10038      DRIVER  c21c476e2ccc41af550b5d858d22aaac4ffc88745a1700...   
4    O10039      DRIVER  eb390a4c8e114c69488f5fb8a097fe629f5a92fd528cf4...   

      RD_NO VEHICLE_ID              CRASH_DATE SEAT_NO     CITY STATE ZIPCODE  \
0  JC451435     834816  09/28/2019 03:30:00 AM     NaN  CHICAGO    IL   60651   
1  JD208731     827212  04/13/2020 10:50:00 PM     NaN  CHICAGO    IL   60620   
2  HY484534       9579  11/01/2015 05:00:00 AM     NaN      NaN   NaN     NaN   
3  HY484750       9598  11/01/2015 08:00:00 AM     NaN      NaN   NaN     NaN   
4  HY484778       9600  11/01/2015 10:15:00 AM     NaN      NaN   NaN     NaN   

   ... EMS_RUN_NO     DRIVER_ACTION DRIVER_VISION PHYSICAL_CONDITION  \
0  ...        NaN           UNKNOWN       UNKNOWN            UNKNOWN   
1  ...        NaN              NONE  NOT OBSCURED             NORMAL   
2  ...        NaN  IMPROPER BACKING       UNKNOWN            UNKNOWN   
3  ...        NaN           UNKNOWN       UNKNOWN            UNKNOWN   
4  ...        NaN           UNKNOWN       UNKNOWN            UNKNOWN   

  PEDPEDAL_ACTION PEDPEDAL_VISIBILITY PEDPEDAL_LOCATION        BAC_RESULT  \
0             NaN                 NaN               NaN  TEST NOT OFFERED   
1             NaN                 NaN               NaN  TEST NOT OFFERED   
2             NaN                 NaN               NaN  TEST NOT OFFERED   
3             NaN                 NaN               NaN  TEST NOT OFFERED   
4             NaN                 NaN               NaN  TEST NOT OFFERED   

  BAC_RESULT VALUE CELL_PHONE_USE  
0              NaN            NaN  
1              NaN            NaN  
2              NaN            NaN  
3              NaN            NaN  
4              NaN            NaN  

[5 rows x 30 columns]

In [25]:
df_people_raw.columns

Index(['PERSON_ID', 'PERSON_TYPE', 'CRASH_RECORD_ID', 'RD_NO', 'VEHICLE_ID',
       'CRASH_DATE', 'SEAT_NO', 'CITY', 'STATE', 'ZIPCODE', 'SEX', 'AGE',
       'DRIVERS_LICENSE_STATE', 'DRIVERS_LICENSE_CLASS', 'SAFETY_EQUIPMENT',
       'AIRBAG_DEPLOYED', 'EJECTION', 'INJURY_CLASSIFICATION', 'HOSPITAL',
       'EMS_AGENCY', 'EMS_RUN_NO', 'DRIVER_ACTION', 'DRIVER_VISION',
       'PHYSICAL_CONDITION', 'PEDPEDAL_ACTION', 'PEDPEDAL_VISIBILITY',
       'PEDPEDAL_LOCATION', 'BAC_RESULT', 'BAC_RESULT VALUE',
       'CELL_PHONE_USE'],
      dtype='object')

With the people dataset, we will be looking at the personal characterstics of the person and action he/she took to classify the car crash. With this in mind, we can remove the following variables.

In [26]:
columns_to_delete_people = ['PERSON_ID', 'RD_NO', 'CRASH_DATE', 'SEAT_NO', 'CITY', 'STATE', 'ZIPCODE',
       'DRIVERS_LICENSE_STATE', 'DRIVERS_LICENSE_CLASS', 'SAFETY_EQUIPMENT', 'AIRBAG_DEPLOYED', 'EJECTION', 
       'INJURY_CLASSIFICATION', 'HOSPITAL', 'EMS_AGENCY', 'EMS_RUN_NO','PEDPEDAL_LOCATION']

In [27]:
# Drop irrelevant columns

df_people_clean = df_people_raw.drop(columns_to_delete_people, axis=1).copy()

In [28]:
# Check for missing values

df_people_clean.isnull().sum()

PERSON_TYPE                  0
CRASH_RECORD_ID              0
VEHICLE_ID               24298
SEX                      18549
AGE                     352917
DRIVER_ACTION           252940
DRIVER_VISION           253279
PHYSICAL_CONDITION      252247
PEDPEDAL_ACTION        1204023
PEDPEDAL_VISIBILITY    1204073
BAC_RESULT              251851
BAC_RESULT VALUE       1225378
CELL_PHONE_USE         1225765
dtype: int64

In [29]:
# Check the total count of the dataset

df_people_clean.shape

(1226922, 13)

We have majority of values missing in the following columns:
- PEPPEDAL_ACTION,
- PEDPEDAL_VISIBILITY,
- BAC_RESULT_VALUE, and 
- CELL_PHONE_USE

We will drop these columns.

In [30]:
df_people_clean.drop(['PEDPEDAL_ACTION','PEDPEDAL_VISIBILITY', 'BAC_RESULT VALUE', 
                      'CELL_PHONE_USE'], axis=1, inplace=True)

In [31]:
df_people_clean.isnull().sum()

PERSON_TYPE                0
CRASH_RECORD_ID            0
VEHICLE_ID             24298
SEX                    18549
AGE                   352917
DRIVER_ACTION         252940
DRIVER_VISION         253279
PHYSICAL_CONDITION    252247
BAC_RESULT            251851
dtype: int64

The amount of missing values are about 20% of the dataset so we cannot drop them. We will look deeper into what kind of information these columns hold. 

In [32]:
# Loop through each of the columns and print it's unique values

for col in df_people_clean.columns:
    print('\n' + col + '\n')
    print(df_people_clean[col].value_counts())


PERSON_TYPE

DRIVER                 952473
PASSENGER              250616
PEDESTRIAN              14008
BICYCLE                  8549
NON-MOTOR VEHICLE        1055
NON-CONTACT VEHICLE       221
Name: PERSON_TYPE, dtype: int64

CRASH_RECORD_ID

31ecf6862c691ff12d3856213b902c146b07337b42a5692e3a176a66d684d221028bb5118ef6d67a313bcaed9e97bee1855cb1f5e8650f49e8dc17663475a1ee    61
13026c7fb51566d9ca487a093e38c6f5621c2ec25be48c306b6574983b61daeee589524b96bb2bfe66ddd0f695c8d2bf3ab0297558528e9c7a70363c763d6bd1    50
1829f52c1281a0396ef94692331b3dc530bc4be5a54cd55e94c24a5e5e49b800fbcf9f24dabe4c8277c8964ad05aadc89e90fd94021959d6dff5fad55480d595    46
c727dc759107cf17b2e8141149347128bb4bc26b026c7805562206c7c5761c543dd7cc0e47fc11379455a2ecbb2847c3d1744d6feb78f276d9a457e9beeb6121    45
60fbcef1942e23e9d93ba3fd5dd7f1da2c5b1e4bcc30731363a0f9ed822b5971a38d109cace85875dc2a7321d8a28ba40f24548d9ddd728f600af3b08c036a70    44
                                                                                 

Looking at the details above:

- We are only interested in the DRIVER for PERSON_TYPE. We will slice the rest out of the dataset.
- 'SEX - X' and negative 'AGE' values does not explain much about the person. We will slice them out of the dataset.
- DRIVER_ACTION seems more of a target variable rather than a predictor. We will drop this column.
- BAC_RESULT has mostly TEST NOT OFFERED. The rest of the data will not be able to add much weight for this predictor. We will drop this as well.

In [33]:
# Drop the irrelevant columns first

df_people_clean.drop(['DRIVER_ACTION', 'BAC_RESULT'], axis=1, inplace=True)

In [34]:
# Slice the rest of the columns as mentioned

df_people_clean = df_people_clean[(df_people_clean['PERSON_TYPE'] == 'DRIVER') & 
                                  (df_people_clean['SEX'] != 'X') & 
                                  (df_people_clean['AGE'] > '0')]

In [35]:
df_people_clean.isnull().sum()

PERSON_TYPE             0
CRASH_RECORD_ID         0
VEHICLE_ID            374
SEX                     0
AGE                     0
DRIVER_VISION           0
PHYSICAL_CONDITION      0
dtype: int64

The 374 missing values in VEHICLE_ID is minimal. We will drop these values.

In [36]:
# Remove the missing values

df_people_clean.dropna(axis=0, inplace=True)

In [37]:
# Check the final count of the dataset

df_people_clean.shape

(693111, 7)

Now, that all the datasets are cleaned, we will merge them together.

### Merge all datasets

We will start the merge with the 'Vehicles' and 'People' dataset first. Since these are instances reported in a crash where a vehicle was involved, we will use the VEHICLE_ID as the unique key to merge them.

In [38]:
df_vehicles_people = pd.merge(df_vehicles_clean, df_people_clean, on=['VEHICLE_ID'], how = 'inner')

In [39]:
df_vehicles_people.head()

UNIT_NO UNIT_TYPE VEHICLE_ID VEHICLE_DEFECT VEHICLE_TYPE VEHICLE_USE  \
0       1    DRIVER     796949           NONE    PASSENGER    PERSONAL   
1       1    DRIVER     834816        UNKNOWN    PASSENGER    PERSONAL   
2       2    DRIVER     827212           NONE    PASSENGER    PERSONAL   
3       1    DRIVER     792247           NONE    PASSENGER    PERSONAL   
4       2    DRIVER     792252           NONE    PASSENGER    PERSONAL   

  TRAVEL_DIRECTION        MANEUVER FIRST_CONTACT_POINT PERSON_TYPE  \
0                N  STRAIGHT AHEAD               FRONT      DRIVER   
1                N  STRAIGHT AHEAD               FRONT      DRIVER   
2                W  STRAIGHT AHEAD         FRONT-RIGHT      DRIVER   
3                W  STRAIGHT AHEAD               OTHER      DRIVER   
4                S  STRAIGHT AHEAD                ROOF      DRIVER   

                                     CRASH_RECORD_ID SEX AGE DRIVER_VISION  \
0  24ddf9fd8542199d832e1c223cc474e5601b356f1d77a6...   M  35  NOT OBSCURED   
1  81dc0de2ed92aa62baccab641fa377be7feb1cc47e6554...   M  25       UNKNOWN   
2  af84fb5c8d996fcd3aefd36593c3a02e6e7509eeb27568...   M  37  NOT OBSCURED   
3  ecf9e646f89a73ed3df0f373e4fbd18e91f5a595004019...   F  32  NOT OBSCURED   
4  ecf9e646f89a73ed3df0f373e4fbd18e91f5a595004019...   F  37  NOT OBSCURED   

  PHYSICAL_CONDITION  
0             NORMAL  
1            UNKNOWN  
2             NORMAL  
3             NORMAL  
4             NORMAL

Now we can use the CRASH_RECORD_ID to combine the rest of the details from 'Crashes' dataset.

In [40]:
df_vehicles_people_crashes = pd.merge(df_vehicles_people, df_crashes_clean, on=['CRASH_RECORD_ID'], how='inner')

In [41]:
df_vehicles_people_crashes.head()

UNIT_NO UNIT_TYPE VEHICLE_ID VEHICLE_DEFECT VEHICLE_TYPE VEHICLE_USE  \
0       1    DRIVER     796949           NONE    PASSENGER    PERSONAL   
1       1    DRIVER     834816        UNKNOWN    PASSENGER    PERSONAL   
2       2    DRIVER     827212           NONE    PASSENGER    PERSONAL   
3       1    DRIVER     792247           NONE    PASSENGER    PERSONAL   
4       2    DRIVER     792252           NONE    PASSENGER    PERSONAL   

  TRAVEL_DIRECTION        MANEUVER FIRST_CONTACT_POINT PERSON_TYPE  ...  \
0                N  STRAIGHT AHEAD               FRONT      DRIVER  ...   
1                N  STRAIGHT AHEAD               FRONT      DRIVER  ...   
2                W  STRAIGHT AHEAD         FRONT-RIGHT      DRIVER  ...   
3                W  STRAIGHT AHEAD               OTHER      DRIVER  ...   
4                S  STRAIGHT AHEAD                ROOF      DRIVER  ...   

                         CRASH_TYPE         DAMAGE  \
0            NO INJURY / DRIVE AWAY  $501 - $1,500   
1            NO INJURY / DRIVE AWAY    OVER $1,500   
2            NO INJURY / DRIVE AWAY   $500 OR LESS   
3  INJURY AND / OR TOW DUE TO CRASH    OVER $1,500   
4  INJURY AND / OR TOW DUE TO CRASH    OVER $1,500   

        PRIM_CONTRIBUTORY_CAUSE NUM_UNITS CRASH_HOUR  CRASH_DAY_OF_WEEK  \
0           UNABLE TO DETERMINE         1          6                  4   
1           UNABLE TO DETERMINE         3          3                  7   
2   IMPROPER OVERTAKING/PASSING         2         22                  2   
3  DISREGARDING TRAFFIC SIGNALS         2         15                  2   
4  DISREGARDING TRAFFIC SIGNALS         2         15                  2   

  CRASH_MONTH   LATITUDE  LONGITUDE                                  LOCATION  
0           1  41.860250 -87.695575  POINT (-87.695575177986 41.860250253825)  
1           9  41.900043 -87.755577  POINT (-87.755576950444 41.900042872883)  
2           4  41.736044 -87.653404  POINT (-87.653404241798 41.736044089544)  
3           2  41.997755 -87.694830  POINT (-87.694829722211 41.997754789249)  
4           2  41.997755 -87.694830  POINT (-87.694829722211 41.997754789249)  

[5 rows x 35 columns]

In [42]:
# Check all columns included in the final dataset

df_vehicles_people_crashes.columns

Index(['UNIT_NO', 'UNIT_TYPE', 'VEHICLE_ID', 'VEHICLE_DEFECT', 'VEHICLE_TYPE',
       'VEHICLE_USE', 'TRAVEL_DIRECTION', 'MANEUVER', 'FIRST_CONTACT_POINT',
       'PERSON_TYPE', 'CRASH_RECORD_ID', 'SEX', 'AGE', 'DRIVER_VISION',
       'PHYSICAL_CONDITION', 'POSTED_SPEED_LIMIT', 'TRAFFIC_CONTROL_DEVICE',
       'DEVICE_CONDITION', 'WEATHER_CONDITION', 'LIGHTING_CONDITION',
       'FIRST_CRASH_TYPE', 'TRAFFICWAY_TYPE', 'ALIGNMENT',
       'ROADWAY_SURFACE_COND', 'ROAD_DEFECT', 'CRASH_TYPE', 'DAMAGE',
       'PRIM_CONTRIBUTORY_CAUSE', 'NUM_UNITS', 'CRASH_HOUR',
       'CRASH_DAY_OF_WEEK', 'CRASH_MONTH', 'LATITUDE', 'LONGITUDE',
       'LOCATION'],
      dtype='object')

In [43]:
df_vehicles_people_crashes.head()

UNIT_NO UNIT_TYPE VEHICLE_ID VEHICLE_DEFECT VEHICLE_TYPE VEHICLE_USE  \
0       1    DRIVER     796949           NONE    PASSENGER    PERSONAL   
1       1    DRIVER     834816        UNKNOWN    PASSENGER    PERSONAL   
2       2    DRIVER     827212           NONE    PASSENGER    PERSONAL   
3       1    DRIVER     792247           NONE    PASSENGER    PERSONAL   
4       2    DRIVER     792252           NONE    PASSENGER    PERSONAL   

  TRAVEL_DIRECTION        MANEUVER FIRST_CONTACT_POINT PERSON_TYPE  ...  \
0                N  STRAIGHT AHEAD               FRONT      DRIVER  ...   
1                N  STRAIGHT AHEAD               FRONT      DRIVER  ...   
2                W  STRAIGHT AHEAD         FRONT-RIGHT      DRIVER  ...   
3                W  STRAIGHT AHEAD               OTHER      DRIVER  ...   
4                S  STRAIGHT AHEAD                ROOF      DRIVER  ...   

                         CRASH_TYPE         DAMAGE  \
0            NO INJURY / DRIVE AWAY  $501 - $1,500   
1            NO INJURY / DRIVE AWAY    OVER $1,500   
2            NO INJURY / DRIVE AWAY   $500 OR LESS   
3  INJURY AND / OR TOW DUE TO CRASH    OVER $1,500   
4  INJURY AND / OR TOW DUE TO CRASH    OVER $1,500   

        PRIM_CONTRIBUTORY_CAUSE NUM_UNITS CRASH_HOUR  CRASH_DAY_OF_WEEK  \
0           UNABLE TO DETERMINE         1          6                  4   
1           UNABLE TO DETERMINE         3          3                  7   
2   IMPROPER OVERTAKING/PASSING         2         22                  2   
3  DISREGARDING TRAFFIC SIGNALS         2         15                  2   
4  DISREGARDING TRAFFIC SIGNALS         2         15                  2   

  CRASH_MONTH   LATITUDE  LONGITUDE                                  LOCATION  
0           1  41.860250 -87.695575  POINT (-87.695575177986 41.860250253825)  
1           9  41.900043 -87.755577  POINT (-87.755576950444 41.900042872883)  
2           4  41.736044 -87.653404  POINT (-87.653404241798 41.736044089544)  
3           2  41.997755 -87.694830  POINT (-87.694829722211 41.997754789249)  
4           2  41.997755 -87.694830  POINT (-87.694829722211 41.997754789249)  

[5 rows x 35 columns]

In [44]:
# Check the final count of the dataset

df_vehicles_people_crashes.shape

(687278, 35)

Our cleaned dataset has over 680,000 rows with 35 columns. Almost all of the columns are categorical. We will have to do some feature elimination to make sure this fits into a model.

### Feature Engineering

Feature engineering is a process to add, remove, create, or combine different classes to better suit the purpose of the EDA. We will look at both Target and Predictor variables to see if we need to change some values to make it more meaningful or reduce it for modeling.

In [45]:
# Check all columns present in the dataset

df_vehicles_people_crashes.columns

Index(['UNIT_NO', 'UNIT_TYPE', 'VEHICLE_ID', 'VEHICLE_DEFECT', 'VEHICLE_TYPE',
       'VEHICLE_USE', 'TRAVEL_DIRECTION', 'MANEUVER', 'FIRST_CONTACT_POINT',
       'PERSON_TYPE', 'CRASH_RECORD_ID', 'SEX', 'AGE', 'DRIVER_VISION',
       'PHYSICAL_CONDITION', 'POSTED_SPEED_LIMIT', 'TRAFFIC_CONTROL_DEVICE',
       'DEVICE_CONDITION', 'WEATHER_CONDITION', 'LIGHTING_CONDITION',
       'FIRST_CRASH_TYPE', 'TRAFFICWAY_TYPE', 'ALIGNMENT',
       'ROADWAY_SURFACE_COND', 'ROAD_DEFECT', 'CRASH_TYPE', 'DAMAGE',
       'PRIM_CONTRIBUTORY_CAUSE', 'NUM_UNITS', 'CRASH_HOUR',
       'CRASH_DAY_OF_WEEK', 'CRASH_MONTH', 'LATITUDE', 'LONGITUDE',
       'LOCATION'],
      dtype='object')

Let's look into each one of our columns and see how many classes are present and can be modified.

In [46]:
# Loop through each of the columns and print it's unique values

for col in df_vehicles_people_crashes.columns:
    print('\n' + col + '\n')
    print(df_vehicles_people_crashes[col].value_counts())


UNIT_NO

2     347927
1     321706
3      15365
4       1830
5        319
6         83
7         31
8         12
9          3
0          1
10         1
Name: UNIT_NO, dtype: int64

UNIT_TYPE

DRIVER    687278
Name: UNIT_TYPE, dtype: int64

VEHICLE_ID

316605     1
1075946    1
289156     1
1091489    1
356566     1
          ..
253194     1
481761     1
21829      1
69861      1
393730     1
Name: VEHICLE_ID, Length: 687278, dtype: int64

VEHICLE_DEFECT

NONE                465329
UNKNOWN             214583
BRAKES                3134
OTHER                 2579
TIRES                  474
STEERING               465
WHEELS                 236
SUSPENSION             171
ENGINE/MOTOR           112
LIGHTS                  45
FUEL SYSTEM             44
WINDOWS                 38
CARGO                   20
SIGNALS                 15
RESTRAINT SYSTEM        13
TRAILER COUPLING        10
EXHAUST                 10
Name: VEHICLE_DEFECT, dtype: int64

VEHICLE_TYPE

PASSENGER                      

DAYLIGHT                  475805
DARKNESS, LIGHTED ROAD    139300
DARKNESS                   28575
DUSK                       21579
DAWN                       11412
UNKNOWN                    10607
Name: LIGHTING_CONDITION, dtype: int64

FIRST_CRASH_TYPE

REAR END                        214259
TURNING                         130833
SIDESWIPE SAME DIRECTION        130417
ANGLE                            98149
PARKED MOTOR VEHICLE             46385
FIXED OBJECT                     18110
SIDESWIPE OPPOSITE DIRECTION     12089
HEAD ON                           7679
PEDESTRIAN                        7509
REAR TO FRONT                     5790
PEDALCYCLIST                      5206
REAR TO SIDE                      3800
OTHER OBJECT                      3786
OTHER NONCOLLISION                1361
REAR TO REAR                      1335
ANIMAL                             285
OVERTURNED                         252
TRAIN                               33
Name: FIRST_CRASH_TYPE, dtype: int64

TRAF

Looking at the values in each of our columns, we can make the following adjustments.

* UNIT_NO has most values in 1 and 2. Since we are looking at the accident as a whole and not individual vehices, we will drop this column.
* UNIT_TYPE has only one value. We will drop this.
* VEHICLE_ID is not required anymore. We will drop this.
* VEHICLE_DEFECT has majority values as NONE or UNKNOWN. This does not give us much insight so we will drop this.
* VEHICLE_TYPE has majority values for PASSENGER. We are not able to say for sure if the other classes are either passenger or not . We will drop this column.
* VEHICLE_USE has most values as PERSONAL, UNKNOWN/NA and OTHER. We will drop this.
* TRAVEL_DIRECTION does not mean much to our analysis. We will drop this.
* MANEUVER is something we are interested in so we will keep this as is.
* FIRST_CONTACT_POINT is not relevant to our analysis. We will drop this.
* PERSON_TYPE only has one class. We will drop this column.
* CRASH_RECORD_ID is not required anymore. We will drop this.
* SEX and AGE can be left as is.
* FIRST_CRASH_TYPE has very distinct classes. Although irregularly distributed, all of them seem important. We will leave this as is. 
* ALIGNMENT has most values in only one class. We will drop this column.
* PRIM_CONTRIBUTORY_CAUSE also has very distinct classes. We will leave them as is.
* NUM_UNITS has most values in 2, 3 and 1. Since there are not too many classes with enough values, we will drop this.
* CRASH_HOUR, CRASH_DAY_OF_WEEK, CRASH_MONTH, LATITUDE, LONGITUDE, and LOCATION are specific to each accident. we will leave them as is.
* Rest of the columns consist of information that we need for our analysis. We will look at them individually. 

For our target variable, we will use CRASH_TYPE and DAMAGE. Combining these two columns, we will create four new classes where all our input variables will converge.

- NO INJURY / DAMAGE < 1500
- NO INJURY / DAMAGE > 1500
- INJURY / DAMAGE < 1500
- INJURY / DAMAGE > 1500

In [47]:
columns_to_drop = ['UNIT_NO', 'UNIT_TYPE', 'VEHICLE_ID', 'VEHICLE_DEFECT', 'VEHICLE_TYPE',
       'VEHICLE_USE', 'TRAVEL_DIRECTION', 'FIRST_CONTACT_POINT',
       'PERSON_TYPE', 'CRASH_RECORD_ID', 'NUM_UNITS']

In [48]:
# Drop the irrelevant columns and save the data in a new dataframe

df_clean = df_vehicles_people_crashes.drop(columns_to_drop, axis=1).copy()

In [49]:
df_clean.columns

Index(['MANEUVER', 'SEX', 'AGE', 'DRIVER_VISION', 'PHYSICAL_CONDITION',
       'POSTED_SPEED_LIMIT', 'TRAFFIC_CONTROL_DEVICE', 'DEVICE_CONDITION',
       'WEATHER_CONDITION', 'LIGHTING_CONDITION', 'FIRST_CRASH_TYPE',
       'TRAFFICWAY_TYPE', 'ALIGNMENT', 'ROADWAY_SURFACE_COND', 'ROAD_DEFECT',
       'CRASH_TYPE', 'DAMAGE', 'PRIM_CONTRIBUTORY_CAUSE', 'CRASH_HOUR',
       'CRASH_DAY_OF_WEEK', 'CRASH_MONTH', 'LATITUDE', 'LONGITUDE',
       'LOCATION'],
      dtype='object')

In [50]:
df_clean.shape

(687278, 24)

We now have 24 columns. If we create the target variable using two of the columns as explained above, we will end up with 22 columns or predictor variables. The unique values also show unequal distribution of data across each predictor. This can lead to model bias, which attributes more importance to classes that are more in numbers than those that are not, thus giving us incorrect results. To solve this, we will have to look into each of these predictors independently and see if we can combine some categories to make the distribution normal.

We are now down to 24 columns. However, we can see that each of these predictors have multiple classes. It will be very difficult for us to model each and every class present in every predictor.  

### Predictor Variables

#### MANEUVER

In [51]:
df_clean.MANEUVER.value_counts()

STRAIGHT AHEAD                        392414
SLOW/STOP IN TRAFFIC                   76486
TURNING LEFT                           52665
BACKING                                29930
TURNING RIGHT                          28219
UNKNOWN/NA                             16744
CHANGING LANES                         14848
PASSING/OVERTAKING                     14186
OTHER                                  13418
ENTERING TRAFFIC LANE FROM PARKING      9600
STARTING IN TRAFFIC                     5597
MERGING                                 5540
U-TURN                                  4319
AVOIDING VEHICLES/OBJECTS               4214
LEAVING TRAFFIC LANE TO PARK            4008
SKIDDING/CONTROL LOSS                   3640
ENTER FROM DRIVE/ALLEY                  3342
SLOW/STOP - LEFT TURN                   2684
SLOW/STOP - RIGHT TURN                  1650
SLOW/STOP - LOAD/UNLOAD                 1446
NEGOTIATING A CURVE                     1077
DRIVING WRONG WAY                        829
TURNING ON

All values are unique to itself so we will leave this column as is.

#### SEX

In [52]:
df_clean.SEX.value_counts()

M    410779
F    276499
Name: SEX, dtype: int64

There are only two classes here. However, we will have to take note that there are almost double of males than females in our dataset. While interpreting the results, we will have to keep this in consideration.

The count of Males are almost double than that of Females. The model will give higher probablity of an accident to a male than a female, but this is what our data shows. 

#### AGE

In [53]:
df_clean.AGE.value_counts()

25     20676
27     20577
26     20525
28     20015
24     19569
       ...  
107        2
104        2
109        2
110        2
108        2
Name: AGE, Length: 107, dtype: int64

AGE is a numerical value and can be left as is.

#### DRIVER VISION

In [54]:
df_clean.DRIVER_VISION.value_counts()

NOT OBSCURED              470250
UNKNOWN                   195871
OTHER                       7785
MOVING VEHICLES             5404
PARKED VEHICLES             3252
WINDSHIELD (WATER/ICE)      2625
BLINDED - SUNLIGHT          1096
TREES, PLANTS                397
BUILDINGS                    306
BLINDED - HEADLIGHTS          78
HILLCREST                     74
BLOWING MATERIALS             63
EMBANKMENT                    56
SIGNBOARD                     21
Name: DRIVER_VISION, dtype: int64

- NOT OBSCURED has the highest number of values. We will leave them as os.
- UNKNOWN cannot be classified into anything else so we will zero them out.
- Rest of the classes all have very small number of values and are different variations of being obscured. We will group these together as OBSCURED.

In [55]:
df_clean.DRIVER_VISION = df_clean.DRIVER_VISION.map(lambda x: 'OBSCURED' 
                                                         if x not in ['NOT OBSCURED', 
                                                                      'UNKNOWN'] else x)
df_clean.DRIVER_VISION = df_clean.DRIVER_VISION.map(lambda x: 0
                                                         if x not in ['NOT OBSCURED',
                                                                      'OBSCURED'] else x)

In [56]:
df_clean.DRIVER_VISION.value_counts()

NOT OBSCURED    470250
0               195871
OBSCURED         21157
Name: DRIVER_VISION, dtype: int64

#### PHYSICAL CONDITION

In [57]:
df_clean.PHYSICAL_CONDITION.value_counts()

NORMAL                          583584
UNKNOWN                          89038
IMPAIRED - ALCOHOL                3943
REMOVED BY EMS                    2412
FATIGUED/ASLEEP                   2368
OTHER                             1954
EMOTIONAL                         1853
ILLNESS/FAINTED                    864
HAD BEEN DRINKING                  502
IMPAIRED - DRUGS                   460
IMPAIRED - ALCOHOL AND DRUGS       192
MEDICATED                          108
Name: PHYSICAL_CONDITION, dtype: int64

- NORMAL has the highest number of values. We will leave them as is.
- UNKNOWN will be changed to zero.
- Everything else is a version of IMPARIED. We will rename the rest as IMPAIRED.

In [58]:
df_clean.PHYSICAL_CONDITION = df_clean.PHYSICAL_CONDITION.map(lambda x: 'IMPAIRED' 
                                                                   if x not in ['NORMAL', 
                                                                                'UNKNOWN'] else x)
df_clean.PHYSICAL_CONDITION = df_clean.PHYSICAL_CONDITION.map(lambda x: 0
                                                                   if x == 'UNKNOWN' else x)

In [59]:
df_clean.PHYSICAL_CONDITION.value_counts()

NORMAL      583584
0            89038
IMPAIRED     14656
Name: PHYSICAL_CONDITION, dtype: int64

#### POSTED SPEED LIMIT

In [60]:
df_clean.POSTED_SPEED_LIMIT.value_counts()

30    526232
35     52821
25     36525
20     19937
15     16537
10      9956
40      8027
0       7271
45      5065
5       3319
55       731
3        177
50       176
9        105
99        83
39        81
60        39
1         38
24        20
32        20
2         18
34        17
33        15
65        13
11         9
6          8
36         6
14         4
31         4
12         3
7          3
18         2
38         2
22         2
63         2
4          2
70         2
29         2
26         2
49         1
23         1
Name: POSTED_SPEED_LIMIT, dtype: int64

We know that speed limits are usually in multiples of 5. We'll use this logic to group the above variation. The national speed limimt for school area is 15 mph, which is also the lowest speed limit. We will use this as the cutoff.

In [61]:
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 15 if x <= 15 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 20 if 15 < x <= 20 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 25 if 20 < x <= 25 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 30 if 25 < x <= 30 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 35 if 30 < x <= 35 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 40 if 35 < x <= 40 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 45 if 40 < x <= 45 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 50 if 45 < x <= 50 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 55 if 50 < x <= 55 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 60 if 55 < x <= 60 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 65 if 60 < x <= 65 else x)
df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: 70 if 65 < x else x)

In [62]:
df_clean.POSTED_SPEED_LIMIT.value_counts()

30    526236
35     52877
15     37450
25     36548
20     19939
40      8116
45      5065
55       731
50       177
70        85
60        39
65        15
Name: POSTED_SPEED_LIMIT, dtype: int64

In [63]:
type(df_clean.POSTED_SPEED_LIMIT[1])

numpy.int64

We also need these as classes so we will transform them into string values.

In [64]:
# Convert SPEED values to string

df_clean.POSTED_SPEED_LIMIT = df_clean.POSTED_SPEED_LIMIT.map(lambda x: str(x))

#### TRAFFIC CONTROL DEVICE

In [65]:
df_clean.TRAFFIC_CONTROL_DEVICE.value_counts()

NO CONTROLS                 337581
TRAFFIC SIGNAL              239472
STOP SIGN/FLASHER            81014
UNKNOWN                      18824
OTHER                         4061
LANE USE MARKING              1812
YIELD                         1145
OTHER REG. SIGN                663
OTHER WARNING SIGN             581
RAILROAD CROSSING GATE         519
PEDESTRIAN CROSSING SIGN       306
POLICE/FLAGMAN                 301
FLASHING CONTROL SIGNAL        254
DELINEATORS                    237
SCHOOL ZONE                    193
OTHER RAILROAD CROSSING        177
RR CROSSING SIGN                80
NO PASSING                      41
BICYCLE CROSSING SIGN           17
Name: TRAFFIC_CONTROL_DEVICE, dtype: int64

- NO CONTROLS has the highest number of values. We will leave them as is.
- TRAFFIC SIGNAL has the second highest number of values. We will also leave them as is.
- UNKNOWN will be changed to zero.
- Rest can be grouped into one as they all have smaller values and are some form of OTHER CONTROLS.

In [66]:
# Rename the rest as OTHER CONTROLS

df_clean.TRAFFIC_CONTROL_DEVICE = df_clean.TRAFFIC_CONTROL_DEVICE.map(lambda x: 'OTHER CONTROLS'
                                                                           if x not in ['NO CONTROLS', 
                                                                                        'TRAFFIC SIGNAL',
                                                                                        'UNKNOWN'] else x)
df_clean.TRAFFIC_CONTROL_DEVICE = df_clean.TRAFFIC_CONTROL_DEVICE.map(lambda x: 0
                                                                           if x == 'UNKNOWN' else x)

In [67]:
df_clean.TRAFFIC_CONTROL_DEVICE.value_counts()

NO CONTROLS       337581
TRAFFIC SIGNAL    239472
OTHER CONTROLS     91401
0                  18824
Name: TRAFFIC_CONTROL_DEVICE, dtype: int64

#### DEVICE CONDITION

In [68]:
df_clean.DEVICE_CONDITION.value_counts()

NO CONTROLS                 344997
FUNCTIONING PROPERLY        294116
UNKNOWN                      35406
OTHER                         5560
FUNCTIONING IMPROPERLY        4344
NOT FUNCTIONING               2430
WORN REFLECTIVE MATERIAL       329
MISSING                         96
Name: DEVICE_CONDITION, dtype: int64

- NO CONTROLS and FUNCTIONING PROPERLY has the highest number of values. We will leave them as is.
- UNKNOWN, OTHER and MISSING all mean the same thing. We will rename them to zero.
- WORN_RELECTIVE_MATERIAL basically means that the device was functioning properly. We will re-classify them as FUNCTIONING PROPERLY.
* FUNCTIONING IMPROPERLY is as good as NOT FUNCTIONING at all. We will group them together. 

In [69]:
df_clean.DEVICE_CONDITION = df_clean.DEVICE_CONDITION.map(lambda x: 0 
                                                               if x in ['UNKNOWN', 
                                                                        'OTHER',
                                                                        'MISSING'] else x)
df_clean.DEVICE_CONDITION = df_clean.DEVICE_CONDITION.map(lambda x: 'FUNCTIONING' 
                                                               if x in ['FUNCTIONING PROPERLY', 
                                                                        'WORN REFLECTIVE MATERIAL'] else x)
df_clean.DEVICE_CONDITION = df_clean.DEVICE_CONDITION.apply(lambda x: 'NOT FUNCTIONING' 
                                                                 if x in ['FUNCTIONING IMPROPERLY',
                                                                          'NOT FUNCTIONING'] else x)

In [70]:
df_clean.DEVICE_CONDITION.value_counts()

NO CONTROLS        344997
FUNCTIONING        294445
0                   41062
NOT FUNCTIONING      6774
Name: DEVICE_CONDITION, dtype: int64

#### WEATHER CONDITION

In [71]:
df_clean.WEATHER_CONDITION.value_counts()

CLEAR                       553170
RAIN                         64669
SNOW                         24749
CLOUDY/OVERCAST              22669
UNKNOWN                      16687
OTHER                         2088
FOG/SMOKE/HAZE                1064
SLEET/HAIL                     981
FREEZING RAIN/DRIZZLE          877
BLOWING SNOW                   203
SEVERE CROSS WIND GATE         117
BLOWING SAND, SOIL, DIRT         4
Name: WEATHER_CONDITION, dtype: int64

- BLOWING SNOW can be combined with SNOW to create a singular class for when snow is present.
- FREEZING RAIN/DRIZZLE and SLEET/HAIL both occurs during a rain. We will combine them with rain.
- FOG/SMOKE/HAZE, SEVERE CROSS WIND GATE, and BLOWING SAND, SOIL, DIRT, are all unique scenarios but with very low number of instances. We will group them with OTHER to reduce variation and increase number of class observation.
- UNKNOWN will be changed to zero.

In [72]:
df_clean.WEATHER_CONDITION = df_clean.WEATHER_CONDITION.map(lambda x: 'SNOW' 
                                                                 if x == 'BLOWING SNOW' else x)
df_clean.WEATHER_CONDITION = df_clean.WEATHER_CONDITION.map(lambda x: 'RAIN' 
                                                                 if x in ['FREEZING RAIN/DRIZZLE',
                                                                          'SLEET/HAIL'] else x)
df_clean.WEATHER_CONDITION = df_clean.WEATHER_CONDITION.map(lambda x: 'OTHER' 
                                                                 if x in ['FOG/SMOKE/HAZE',
                                                                          'SEVERE CROSS WIND GATE',
                                                                          'BLOWING SAND, SOIL, DIRT'] else x)
df_clean.WEATHER_CONDITION = df_clean.WEATHER_CONDITION.map(lambda x: 0
                                                                 if x == 'UNKNOWN' else x)

In [73]:
df_clean.WEATHER_CONDITION.value_counts()

CLEAR              553170
RAIN                66527
SNOW                24952
CLOUDY/OVERCAST     22669
0                   16687
OTHER                3273
Name: WEATHER_CONDITION, dtype: int64

#### LIGHTING CONDITION

In [74]:
df_clean.LIGHTING_CONDITION.value_counts()

DAYLIGHT                  475805
DARKNESS, LIGHTED ROAD    139300
DARKNESS                   28575
DUSK                       21579
DAWN                       11412
UNKNOWN                    10607
Name: LIGHTING_CONDITION, dtype: int64

- There is two different instances for DARKNESS. We will combine them to be one.
- UNKNOWN will be changed to zero.

In [75]:
df_clean.LIGHTING_CONDITION = df_clean.LIGHTING_CONDITION.map(lambda x: 'DARKNESS' 
                                                                   if x == 'DARKNESS, LIGHTED ROAD' else x)
df_clean.LIGHTING_CONDITION = df_clean.LIGHTING_CONDITION.map(lambda x: 0
                                                                   if x == 'UNKNOWN' else x)

In [76]:
df_clean.LIGHTING_CONDITION.value_counts()

DAYLIGHT    475805
DARKNESS    167875
DUSK         21579
DAWN         11412
0            10607
Name: LIGHTING_CONDITION, dtype: int64

#### TRAFFICWAY TYPE

In [77]:
df_clean.TRAFFICWAY_TYPE.value_counts()

NOT DIVIDED                        314487
DIVIDED - W/MEDIAN (NOT RAISED)    138571
ONE-WAY                             63486
DIVIDED - W/MEDIAN BARRIER          47827
FOUR WAY                            34596
PARKING LOT                         32342
OTHER                               18194
ALLEY                                8006
CENTER TURN LANE                     7235
T-INTERSECTION                       7162
UNKNOWN                              6317
RAMP                                 2430
DRIVEWAY                             1868
UNKNOWN INTERSECTION TYPE            1861
FIVE POINT, OR MORE                   886
Y-INTERSECTION                        849
TRAFFIC ROUTE                         627
NOT REPORTED                          284
ROUNDABOUT                            148
L-INTERSECTION                        102
Name: TRAFFICWAY_TYPE, dtype: int64

- All classes related to INTERSECTION can be grouped into one.
- We can combine the DIVIDED variations into one as well.
- NOT REPORTED is also UNKNOWN. We will rename them both to zero.
- Rest is all unique.

In [78]:
df_clean.TRAFFICWAY_TYPE = df_clean.TRAFFICWAY_TYPE.map(lambda x: 'INTERSECTION' 
                                                             if x in ['T-INTERSECTION',
                                                                      'UNKNOWN INTERSECTION TYPE',
                                                                      'Y-INTERSECTION',
                                                                      'L-INTERSECTION'] else x)
df_clean.TRAFFICWAY_TYPE = df_clean.TRAFFICWAY_TYPE.map(lambda x: 'DIVIDED' 
                                                             if x in ['DIVIDED - W/MEDIAN (NOT RAISED)',
                                                                      'DIVIDED - W/MEDIAN BARRIER'] else x)
df_clean.TRAFFICWAY_TYPE = df_clean.TRAFFICWAY_TYPE.map(lambda x: 0
                                                            if x in ['UNKNOWN',
                                                                     'NOT REPORTED'] else x)

In [79]:
df_clean.TRAFFICWAY_TYPE.value_counts()

NOT DIVIDED            314487
DIVIDED                186398
ONE-WAY                 63486
FOUR WAY                34596
PARKING LOT             32342
OTHER                   18194
INTERSECTION             9974
ALLEY                    8006
CENTER TURN LANE         7235
0                        6601
RAMP                     2430
DRIVEWAY                 1868
FIVE POINT, OR MORE       886
TRAFFIC ROUTE             627
ROUNDABOUT                148
Name: TRAFFICWAY_TYPE, dtype: int64

#### ROADWAY SURFACE CONDITION

In [80]:
df_clean.ROADWAY_SURFACE_COND.value_counts()

DRY                523671
WET                 99315
UNKNOWN             33511
SNOW OR SLUSH       23783
ICE                  5252
OTHER                1518
SAND, MUD, DIRT       228
Name: ROADWAY_SURFACE_COND, dtype: int64

- We will combine ICE with SNOW OR SLUSH creating SNOW/SLUSH/ICE since they all occur during a snow.
- SAND, MUD, DIRT can also be comined to OTHER since they all represent a minority group.
- UNKNOWN will be changed to zero.

In [81]:
df_clean.ROADWAY_SURFACE_COND = df_clean.ROADWAY_SURFACE_COND.map(lambda x: 'SNOW/SLUSH/ICE' 
                                                                       if x in ['SNOW OR SLUSH',
                                                                                'ICE'] else x)
df_clean.ROADWAY_SURFACE_COND = df_clean.ROADWAY_SURFACE_COND.map(lambda x: 'OTHER' 
                                                                       if x in ['OTHER',
                                                                                'SAND, MUD, DIRT'] else x)
df_clean.ROADWAY_SURFACE_COND = df_clean.ROADWAY_SURFACE_COND.map(lambda x: 0
                                                                       if x == 'UNKNOWN' else x)

In [82]:
df_clean.ROADWAY_SURFACE_COND.value_counts()

DRY               523671
WET                99315
0                  33511
SNOW/SLUSH/ICE     29035
OTHER               1746
Name: ROADWAY_SURFACE_COND, dtype: int64

#### ROAD DEFECT

In [83]:
df_clean.ROAD_DEFECT.value_counts()

NO DEFECTS           576466
UNKNOWN               97498
RUT, HOLES             4755
OTHER                  3773
WORN SURFACE           2819
SHOULDER DEFECT        1413
DEBRIS ON ROADWAY       554
Name: ROAD_DEFECT, dtype: int64

- Other than NO DEFECT and UNKNOWN, all others seem to be a variation of DEFECTS. We will group them together as DEFECTS.
- UNKNOWN will be changed to zero.

In [84]:
df_clean.ROAD_DEFECT = df_clean.ROAD_DEFECT.map(lambda x: 'DEFECTS' 
                                                     if x not in ['NO DEFECTS', 
                                                                  'UNKNOWN'] else x)
df_clean.ROAD_DEFECT = df_clean.ROAD_DEFECT.map(lambda x: 0
                                                     if x == 'UNKNOWN' else x)

In [85]:
df_clean.ROAD_DEFECT.value_counts()

NO DEFECTS    576466
0              97498
DEFECTS        13314
Name: ROAD_DEFECT, dtype: int64

#### PRIMARY CONTRIBUTORY CAUSE

In [86]:
df_clean.PRIM_CONTRIBUTORY_CAUSE.value_counts()

UNABLE TO DETERMINE                                                                 201806
FAILING TO YIELD RIGHT-OF-WAY                                                        96154
FOLLOWING TOO CLOSELY                                                                93349
IMPROPER OVERTAKING/PASSING                                                          36121
FAILING TO REDUCE SPEED TO AVOID CRASH                                               35824
IMPROPER LANE USAGE                                                                  29477
NOT APPLICABLE                                                                       28613
IMPROPER TURNING/NO SIGNAL                                                           28394
IMPROPER BACKING                                                                     26258
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  19940
DISREGARDING TRAFFIC SIGNALS                                                         17390

- Everything variable points to a unique class of it's own. We will leave them as is.
- UNKNOWN and NOT APPLICABLE can be changed to zero.

In [87]:
df_clean.PRIM_CONTRIBUTORY_CAUSE = df_clean.PRIM_CONTRIBUTORY_CAUSE.map(lambda x: 0
                                                                            if x in ['UNABLE TO DETERMINE',
                                                                                     'NOT APPLICABLE'] else x)

In [88]:
df_clean.PRIM_CONTRIBUTORY_CAUSE.value_counts()

0                                                                                   230419
FAILING TO YIELD RIGHT-OF-WAY                                                        96154
FOLLOWING TOO CLOSELY                                                                93349
IMPROPER OVERTAKING/PASSING                                                          36121
FAILING TO REDUCE SPEED TO AVOID CRASH                                               35824
IMPROPER LANE USAGE                                                                  29477
IMPROPER TURNING/NO SIGNAL                                                           28394
IMPROPER BACKING                                                                     26258
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  19940
DISREGARDING TRAFFIC SIGNALS                                                         17390
WEATHER                                                                              12357

### Target Variables

We will be combining CRASH TYPE and DAMAGE columns to create our target variables.

In [89]:
(df_clean==0).sum()

MANEUVER                        0
SEX                             0
AGE                             0
DRIVER_VISION              195871
PHYSICAL_CONDITION          89038
POSTED_SPEED_LIMIT              0
TRAFFIC_CONTROL_DEVICE      18824
DEVICE_CONDITION            41062
WEATHER_CONDITION           16687
LIGHTING_CONDITION          10607
FIRST_CRASH_TYPE                0
TRAFFICWAY_TYPE              6601
ALIGNMENT                       0
ROADWAY_SURFACE_COND        33511
ROAD_DEFECT                 97498
CRASH_TYPE                      0
DAMAGE                          0
PRIM_CONTRIBUTORY_CAUSE    230419
CRASH_HOUR                  11096
CRASH_DAY_OF_WEEK               0
CRASH_MONTH                     0
LATITUDE                     4669
LONGITUDE                    4669
LOCATION                     4609
dtype: int64

#### CRASH TYPE

In [90]:
df_clean.CRASH_TYPE.value_counts()

NO INJURY / DRIVE AWAY              490967
INJURY AND / OR TOW DUE TO CRASH    196311
Name: CRASH_TYPE, dtype: int64

We will renname these two columns as INJURY and NO INJURY.

In [91]:
df_clean.CRASH_TYPE = df_clean.CRASH_TYPE.map(lambda x: 'NO INJURY'
                                                   if x == 'NO INJURY / DRIVE AWAY'
                                                         else 'INJURY')

In [92]:
df_clean.CRASH_TYPE.value_counts()

NO INJURY    490967
INJURY       196311
Name: CRASH_TYPE, dtype: int64

#### DAMAGE

In [93]:
df_clean.DAMAGE.value_counts()

OVER $1,500      423110
$501 - $1,500    190521
$500 OR LESS      73647
Name: DAMAGE, dtype: int64

For this, we will create two classes as OVER 1500 and UNDER 1500.

In [94]:
df_clean.DAMAGE = df_clean.DAMAGE.map(lambda x: 'DAMAGE UNDER $1,500' 
                                           if x != 'OVER $1,500' else 'DAMAGE OVER $1,500')

In [95]:
df_clean.DAMAGE.value_counts()

DAMAGE OVER $1,500     423110
DAMAGE UNDER $1,500    264168
Name: DAMAGE, dtype: int64

In [96]:
# Let's combine these two columns to form a single target column

df_clean['TARGET'] = df_clean.CRASH_TYPE + ' & ' + df_clean.DAMAGE

In [97]:
df_clean.TARGET.value_counts()

NO INJURY & DAMAGE OVER $1,500     258320
NO INJURY & DAMAGE UNDER $1,500    232647
INJURY & DAMAGE OVER $1,500        164790
INJURY & DAMAGE UNDER $1,500        31521
Name: TARGET, dtype: int64

Our dataset is finally ready for exploration and modeling. However, we will continue the rest of this project in a new notebook. This will keep our work clean and organized for quick reference.

In [98]:
# Export the final dataset

df_clean.to_csv(r'Datasets/final_dataset.csv', index=False)